File: ServiceBooking.Api/Models/Booking.cs
namespace ServiceBooking.Api.Models;

public class Booking
{
    public int Id { get; set; }

    public int CustomerId { get; set; }

    public int ProviderId { get; set; }

    public int ProviderServiceId { get; set; }

    public DateTime BookingDate { get; set; }

    public BookingStatus Status { get; set; } = BookingStatus.Pending;

    public string? Note { get; set; }

    public decimal TotalPrice { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    // Navigation properties
    public Customer Customer { get; set; } = null!;

    public Provider Provider { get; set; } = null!;

    public ProviderService ProviderService { get; set; } = null!;
}

File: ServiceBooking.Client/src/app/services/bookingservice/booking.ts
import { HttpClient } from '@angular/common/http';
import { inject, Service } from '@angular/core';
import { Observable } from 'rxjs';

export interface BookingItem {
    id: number;
    customerId: number;
    providerId: number;
    providerServiceId: number;
    serviceName: string;
    providerName: string;
    bookingDate: string;
    status: string;
    note?: string;
    totalPrice: number;
    createdAt: string;
}

@Service()
export class Booking {
    private baseUrl = 'http://localhost:5128/api/booking';

    private http = inject(HttpClient);

    getBooking(): Observable<BookingItem[]> {
        const user = localStorage.getItem('user');
        const role = user ? JSON.parse(user).role : '';
        const path = role === 'Provider'
            ? 'provider/bookings'
            : 'customer/bookings';

        return this.http.get<BookingItem[]>(
            `${this.baseUrl}/${path}`
        );
    }
    cancelBooking(id: number) {
        return this.http.patch(
            `${this.baseUrl}/${id}/cancel`,
            {}
        );
    }
}
